File: android/src/test/java/com/audiobrowser/util/PlayingStateFactoryTest.kt
package com.audiobrowser.util

import com.margelo.nitro.audiobrowser.PlaybackState
import org.junit.Assert.assertEquals
import org.junit.Test

/**
 * Truth table for the PlayingState derivation — the single transition function behind
 * `onPlaybackPlayingState` (see Player.refreshPlayingState, its only production caller path).
 */
class PlayingStateFactoryTest {

  @Test
  fun `playWhenReady false is never playing nor buffering`() {
    for (state in PlaybackState.entries) {
      val derived = PlayingStateFactory.derive(playWhenReady = false, playbackState = state)
      assertEquals("playing for $state", false, derived.playing)
      assertEquals("buffering for $state", false, derived.buffering)
    }
  }

  @Test
  fun `playWhenReady true plays except in terminal or empty states`() {
    val notPlaying = setOf(PlaybackState.ERROR, PlaybackState.ENDED, PlaybackState.NONE)
    for (state in PlaybackState.entries) {
      val derived = PlayingStateFactory.derive(playWhenReady = true, playbackState = state)
      assertEquals("playing for $state", state !in notPlaying, derived.playing)
    }
  }

  @Test
  fun `buffering only while loading or rebuffering with playWhenReady`() {
    val buffering = setOf(PlaybackState.LOADING, PlaybackState.BUFFERING)
    for (state in PlaybackState.entries) {
      val derived = PlayingStateFactory.derive(playWhenReady = true, playbackState = state)
      assertEquals("buffering for $state", state in buffering, derived.buffering)
    }
  }
}

File: src/web/TrackPlayer/PlayingStateFactory.test.ts
import { describe, expect, it } from 'vitest'
import type { PlaybackState } from '../../features'
import { derivePlayingState } from './PlayingStateFactory'

// Ported from the native truth table (android PlayingStateFactoryTest.kt /
// ios PlayingStateManager) so web derives PlayingState identically.
const ALL_STATES: PlaybackState[] = [
  'none',
  'ready',
  'playing',
  'paused',
  'stopped',
  'loading',
  'buffering',
  'error',
  'ended'
]

describe('derivePlayingState', () => {
  it('playWhenReady false is never playing nor buffering', () => {
    for (const state of ALL_STATES) {
      expect(derivePlayingState(false, state)).toEqual({
        playing: false,
        buffering: false
      })
    }
  })

  it('playWhenReady true plays except in terminal or empty states', () => {
    const notPlaying = new Set<PlaybackState>(['error', 'ended', 'none'])
    for (const state of ALL_STATES) {
      expect(derivePlayingState(true, state).playing).toBe(!notPlaying.has(state))
    }
  })

  it('buffers only while loading or rebuffering with playWhenReady', () => {
    const buffering = new Set<PlaybackState>(['loading', 'buffering'])
    for (const state of ALL_STATES) {
      expect(derivePlayingState(true, state).buffering).toBe(
        buffering.has(state)
      )
    }
  })
})
